enhancements/olm/marketplace-api-deprecation.md

```diff
@@ -0,0 +1,107 @@
+
+## Summary
+
+This enhancement covers the remaining steps needed to ensure that everything the marketplace-operator APIs have been providing is covered by other features within OLM or OpenShift.
+
+## Open Questions
+
+1. Do we wish to block an upgrade of a cluster if we can detect OperatorSource/CatalogSourceConfig usage that would be affected by the deprecation? (i.e. non-default catalogs using these features).
+2. Are there CVO limitations to be aware of? What happens if we remove a ClusterOperator instance entirely during an upgrade?
+
+## Motivation
+
+In OpenShift 4.1+, the marketplace-operator is responsible for pulling repository data into a cluster from an external source (Quay.io Appregistry) for use by OLM.
+
+As of OpenShift 4.4, alternatives for pulling external data have been built into OLM. The appregistry protocol itself is largely replaced by other efforts in OCI, and we cannot continue to rely on it long-term.
+
+### Goals
+
+- Remove support for appregistry-based storage for operator data
+- Remove the on-cluster components that support it
+- Continue to support configuration for OperatorHub
+
+### Non-Goals
+
+- Preserve non-default OperatorSource and CatalogSourceConfig data
+
+## Proposal
+
+### Removal of the Marketplace APIs
+
+Remove the OperatorSource and CatalogSourceConfig CRDs.
+
+Migrate the default OperatorSources to CatalogSources.
+
+Remove the reconcilation of the OperatorSource and CatalogSourceConfig APIs either by removing the marketplace operator entirely or by removing the relevant portions of it (if it is being retained to manage the OperatorHub API, see below).
+
+### Migration of the OperatorHub Config API
+
+The OperatorHub config API is used to configure which catalogs are available by default in a cluster, and is an important configuration point for disconnected clusters.
+
+This API is currently managed by the marketplace operator and needs to continue to be supported. This could be by continuing to support the marketplace operator as an "operatorhub config operator", by migrating the functionality into OLM, or into a new operator in the OLM repo. The decision here will also determine if we will continue to update a separate ClusterOperator object in the marketplace namespace.
+
+The config API also currently configures OperatorSources, and needs to be changed to configure CatalogSources.
+
+### Risks and Mitigations
+
+It is possible that some customers are heavily using the appregistry-based catalogs and marketplace APIs. Because we believe this use to be minimal at best, we have not provided an automated migration plan for those cases.
+
+If we determine that uninterrupted use of appregistry is desired or required, this proposal will need to include migration steps for all OperatorSource and CatalogSourceConfig APIs in-use in a cluster.
+
+Otherwise, a manual guide for migrating custom catalogs should be straightforward to follow.
```

Comment:
Would it be valuable to provide tooling (via `opm`) to generate an index image from an existing app-registry namespace?

Reply:
`oc` can already do [something similar](https://github.com/operator-framework/olm-book/blob/master/docs/snapshot-appr-registry.md#how-do-i-snapshot-a-quay-appregistry-operator-catalog).